compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -1698,15 +1708,36 @@ impl<'hir> LoweringContext<'_, 'hir> {
         let (iter_pat, iter_pat_nid) =
             self.pat_ident_binding_mode(head_span, iter, hir::BindingAnnotation::MUT);
 
-        // `match Iterator::next(&mut iter) { ... }`
         let match_expr = {
             let iter = self.expr_ident(head_span, iter, iter_pat_nid);
-            let ref_mut_iter = self.expr_mut_addr_of(head_span, iter);
-            let next_expr = self.expr_call_lang_item_fn(
-                head_span,
-                hir::LangItem::IteratorNext,
-                arena_vec![self; ref_mut_iter],
-            );
+            let next_expr = match loop_kind {
+                ForLoopKind::For => {
+                    // `match Iterator::next(&mut iter) { ... }`
+                    let ref_mut_iter = self.expr_mut_addr_of(head_span, iter);
+                    self.expr_call_lang_item_fn(
+                        head_span,
+                        hir::LangItem::IteratorNext,
+                        arena_vec![self; ref_mut_iter],
+                    )
+                }
+                ForLoopKind::ForAwait => {
+                    // `async_iter_next(unsafe { Pin::new_unchecked(&mut iter) }).await`
+                    let future = self.expr_mut_addr_of(head_span, iter);
+                    let future = self.arena.alloc(self.expr_call_lang_item_fn_mut(
+                        head_span,
+                        hir::LangItem::PinNewUnchecked,
+                        arena_vec![self; future],
+                    ));
+                    let future = self.expr_unsafe(future);
+                    let future = self.expr_call_lang_item_fn(
+                        head_span,
+                        hir::LangItem::AsyncIterNext,
```

Comment:
You should not need to call `async_iter_next` at all, actually. I think you can just call `poll_next` in a loop like `await` does, perhaps even sharing logic. This doesn't necessarily block this PR, since I could also just implement this myself and it's an implementation detail at best.

Reply:
I think it's cleaner not to have `async_iter_next` at all, so I'll go ahead and make that change in this PR. We're waiting on lang team approval anyway, so there's no rush.